Updated to latest version of vue.js and converted to use @Component and @Props, etc.

jdiehl opened the pull request:
Thanks for this great starter project.

In this pull request, I updated the project to work with an unmodified version of Vue by converting it to use `@Component` from `vue-class-component` and `@Prop` from `vue-property-decorator`. I also did some modifiationcs to make better use of Visual Studio Code's build-in compiler functions.

msftclas commented:
@__jdiehl__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

msftclas commented:
__@jdiehl__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

sylvainpolletvillard commented on src/components/TodoApp.vue:
```diff
@@ -22,122 +22,6 @@
 </template>
 
 <script lang="ts">
```
You can just write `<script lang="ts" src="./TodoApp">`

sylvainpolletvillard commented on src/components/TodoApp.vue:
```diff
@@ -22,122 +22,6 @@
-  return !todo.completed;
-}
+import TodoApp from './TodoApp';
+export default TodoApp;
```
You can just write <script lang="ts" src="./TodoApp">

jdiehl commented:
I didn't know, you could do that... That's awesome! Fixed.

DanielRosenwasser commented:
Hey there, I think the goal we had with this starter (while it is a work-in-progress) is that we are trying to stick with plain-old Vue.

While I personally think decorators can be awesome, and appreciate the flexibility that Vue has in how you can use it, the way I've come to think about this is that a TypeScript Vue project shouldn't look drastically different from a canonical JavaScript Vue project. We actually put a lot of work into TypeScript 2.3 to accurately type libraries like Vue, Ember, and others so that decorators wouldn't have to be **the** way TypeScript users write Vue.

That said, I would love to hear your thoughts. I'm sure a lot of people aren't hung up on *how* they write Vue as long as they *can* write Vue.

peterblazejewicz commented:
@jdiehl Do you mind updating README.md as well as part of this PR? It depicts the very use-case why custom build is used
Thanks!

sylvainpolletvillard commented:
@DanielRosenwasser One could argue that using `.vue` single file components is already different from a canonical JavaScript Vue project. But `vue-loader` and `vue-class-component` are both maintained by the core team and you can find them on the official vuejs.org docs.

In particular, Class-Style Vue Components with `vue-class-component` are [on the official website under the TypeScript support section](https://vuejs.org/v2/guide/typescript.html#Class-Style-Vue-Components), so I think it is pretty clear that Vue encourages TypeScript users to use this style.

Note also the advantages about type inference that are explained in the docs.

jdiehl commented:
I appreciate your efforts in keeping TypeScript projects as similar as possible to plain vanilla Vue projects and I am sure that this will help many experienced Vue programmers. However, I believe starter projects like this also have the opportunity (and maybe the responsibility) to teach programmers new and easier ways of building apps. Whether the `@Component`-approach falls under that category is not my decision to make.

jdiehl commented:
@peterblazejewicz I'll be happy to work on the Readme but I think we should resolve the issue raised by Daniel first.

octref commented:
Don't underestimate the cost of migration to decorators. One of the major factor in TS's popularity is its low migration cost. Just rename .js to .ts and you get better type check, IntelliSense, etc.

Now it's possible to do that migration by just changing `<script>` to `<script lang="ts">` and wrapping the export object in `Vue.extend`. That's much better than rewriting your whole app with decorators.

Also, the updated type definition can also be used to enhance the editor support for existing Vue projects with js, and the only thing people have to do is to update their vue dependency.

I suggest that you to fork this project or create your own starter template for TypeScript + Vue.

DanielRosenwasser commented:
> @DanielRosenwasser One could argue that using `.vue` single file components is already different from a canonical JavaScript Vue project.

Well that's why I show an ordinary component and then showed a `.vue` file. I think it's fair to say that at this point, `.vue` files are pretty common at this point.

In fact, you can see [there's a relatively small gap](https://npm-stat.com/charts.html?package=vue&package=vue-loader&from=2016-05-13&to=2017-05-13) between vue and vue-loader npm downloads (a smaller margin than vuex and vue-router!). While that is not necessarily the most in-depth analysis, it gives a decent picture.

For what it's worth, I think that having a section about using classes with decorators in Vue would be really useful as well. We might talk about why one might use them over the current approach since there _are_ caveats to the declaration files we've authored when members circularly reference each other.

jdiehl commented:
I have update the readme and fixed several consistency issues. If you do not want to merge this pull request, I would be happy to host it on my fork and would appreciate a link from your readme.

cr0cK commented:
What's about the status of this PR?

It'd be great if this project could be updated with the latest changes of vue 2.5.

jdiehl commented:
This pull request is about decorators. I noticed that the example has significantly changed from the time I created the pull request, so I updated my changes to reflect the current project. I also added the decorators example as an optional part, so everyone can choose how they want to work with Vue and Typescript.

DanielRosenwasser commented on src/components/HelloDecorator.vue:
```diff
@@ -0,0 +1,39 @@
+<!-- src/components/Hello.vue -->
```
Update the comment here, and leave a note about how this is an alternative way to declare the component.

DanielRosenwasser commented on README.md:
```diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
 
+# Use Decorators to define a component
+
+Components can also be defined using [Decorators](https://www.typescriptlang.org/docs/handbook/decorators.html). With the help of two additional packages ([vue-class-component](https://github.com/vuejs/vue-class-component) and [vue-property-decorator](https://github.com/kaorun343/vue-property-decorator)), the definition of our component can be simplified as following:
```
Lowercase the `D` in decorators, use 1 line per sentence, and avoid using the word "simplified" (e.g. written as follows).

DanielRosenwasser commented on README.md:
````diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
+```ts
+@Component
+export default class HelloDecorator extends Vue {
+    @Prop() public name: string;
````
Omit the `public` modifier

DanielRosenwasser commented on README.md:
```diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
+
+Properties are defined by prefixing instance variables with the `@Prop()` decorator from the `vue-property-decorator` package.
+
+Regular instance variables, such as `enthusiasm` in our example, are automatically made available for data binding to the template. Note that all variables must be set to a value other than `undefined` for the binding to work.
```
One sentence per line

DanielRosenwasser replied:
Make it explicit that it's as if they would have declared these properties in the `data` property in Vue's options.

DanielRosenwasser commented on README.md:
````diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
+
+Components can also be defined using [Decorators](https://www.typescriptlang.org/docs/handbook/decorators.html). With the help of two additional packages ([vue-class-component](https://github.com/vuejs/vue-class-component) and [vue-property-decorator](https://github.com/kaorun343/vue-property-decorator)), the definition of our component can be simplified as following:
+
+```ts
````
Make sure that this includes the imports.

DanielRosenwasser commented on README.md:
```diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
+
+Regular instance variables, such as `enthusiasm` in our example, are automatically made available for data binding to the template. Note that all variables must be set to a value other than `undefined` for the binding to work.
+
+Similarly, all instance methods are automatically made available for the template.
```
all instance methods like `increment`

join these sentences to the same paragraph

DanielRosenwasser commented on README.md:
```diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
+
+Similarly, all instance methods are automatically made available for the template.
+
+Finally, the computed properties are implemented as `get` properties.
```
> Finally, computed properties are implemented as `get` accessors.

DanielRosenwasser commented on README.md:
````diff
@@ -442,6 +442,40 @@ Notice a few things about our single-file component:
+}
+```
+
+Instead of using `Vue.extend` to define our component, we create a TypeScript class extending `Vue` and prefix it with the `@Component` decorator from the `vue-class-component` package.
````
If you make your imports explicit, you don't have to mention where the decorators originate from.

jdiehl commented:
@DanielRosenwasser: Thanks for the feedback. I hope the changes address all of your issues.

DanielRosenwasser commented:
@jdiehl there seems to be a huge diff based on whitespace. Can you revert the whitespace changes and amend the commit?

jdiehl commented:
Whatever happened there... It is fixed now.

I appreciate you double-checking everything - should have seen that one myself...

DanielRosenwasser commented:
@jdiehl, I apologize; I meant to send out some changes that I guess got backed out on `README.md` as a PR, but accidentally committed directly to your branch. Let me know if the changes look good and I'll pull it in!

jdiehl commented:
@DanielRosenwasser Looks good.

DanielRosenwasser commented:
Thanks so much!!

sylvainpolletvillard commented:
This is a huge DX improvement here, bravo !

cr0cK commented:
Thank you guys!
